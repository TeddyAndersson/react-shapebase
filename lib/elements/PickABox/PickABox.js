import React from 'react';
import _ from 'lodash';
import styled, { css } from 'styled-components'
import PropTypes from 'prop-types';
import Icon from '../Icon';
import SquareShape from '../Shape/SquareShape';
import Paragraph from '../Paragraph';

class Element extends React.Component{
  constructor(props){
    super(props);
    this.state = {
      activeOption: this.props.initialOption,
      alphabet: [...Array(26)].map((q,w)=>String.fromCharCode(w+97))
    }
  }

  handleClick = (id) => {
    this.setState({activeOption: id})

    _.invoke(this.props, 'onClick', id, this.props)
  }

  render(){
    const {
        className,
        options,
    } = this.props

    return (
        <ul className={className}>
          { _.values(options).map(option => (
            // Replace <li/> with a component for better state handling and generic code
                <li
                    key={option.id}
                    onClick={() => this.handleClick( option.id )}
                >
                    <SquareShape
                        background={`${ option.id === this.state.activeOption ? option.background : 'rgba(0,0,0,0.1)'}`}
                        raised={option.id === this.state.activeOption ? 1 : 0}
                    >
                        <input
                        type="hidden"
                        name={option.name}
                        value={option.value}
                        autoComplete="off"
                        />

                        <div>
                            {
                                option.icon ? <Icon name={option.icon} /> : <Paragraph size={1.5} content={this.state.alphabet[option.index-1]}/>
                            }
                        </div>

                        <div>
                            <span>
                                {option.label}
                            </span>
                        </div>
                    </SquareShape>
                </li>
            ))}
        </ul>
    )
  };
};

Element.propTypes = {

    /**
      * The children of the component.
      */
    children: PropTypes.node,

    /**
      * Additonal classes.
      */
    className: PropTypes.string,

    /**
     * Description of the intreval the range.
     */
    description: PropTypes.bool,

    /**
     * Description of the intreval the range.
     */
    initialOption: PropTypes.string,

    /**
     * The range of the grades options 1-5 or 1-10
     */
    options: PropTypes.arrayOf(PropTypes.shape(
        {
            value: PropTypes.string,
            label: PropTypes.string,
            name: PropTypes.string,
            index: PropTypes.string,
            id: PropTypes.string,
            icon: PropTypes.string,
        }
    ))
}

Element.defaultProps = {
    initialOption: undefined
}

const PickABox = styled(Element)`
    display: flex;
    padding: 0;
    list-style: none;
    margin: 0;
    font-weight: bold;
    font-size: 1em;
    justify-content: space-between;
    
    li {
        margin-bottom: 0.4em;
        div {
            display: flex;
            flex-direction: column;
            justify-content: center;
            align-items: center;
            transition: all 0.3s ease
            height: 6em;
            width: 6em;

            div {
                text-transform: capitalize;
                flex: 6;
            }
            
            div:last-child {
                flex: 1;
                padding: 0.4em;
                font-size: 0.7em;
            }
        }
    }

    li:last-child {
        margin: 0;
    }
    
    ${props => props.range === 10 && css`
        font-size: 0.8em;
    `}
`;

export default PickABox;