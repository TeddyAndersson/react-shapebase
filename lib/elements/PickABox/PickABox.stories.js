import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions/dist/preview';
import Grid from '../Grid';
import PickABox from './PickABox';

const lorem = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."

storiesOf('Pick a box', module)
  .add('Pick a box', () => (
    <Grid columns={1} rowGap={0.5}>
      <PickABox
        onClick={action('Pick a box')}
        options={[
            {
                id: '1',
                name: 'Man',
                value: '1',
                label: 'Man',
                icon: 'mars',
              },
              {
                id: '2',
                name: 'Kvinna',
                value: '2',
                label: 'Kvinna',
                icon: 'venus',
              },
              {
                id: '3',
                name: 'Annat',
                value: '3',
                label: 'Annat',
                icon: 'transgender',
              },
        ]}
      />
      <PickABox
        onClick={action('Pick a box')}
        options={[
            {
                id: '1',
                name: 'Man',
                value: '1',
                label: 'Man',
              },
              {
                id: '2',
                name: 'Kvinna',
                value: '2',
                label: 'Kvinna',
              },
              {
                id: '3',
                name: 'Annat',
                value: '3',
                label: 'Annat',
              },
        ]}
      />
    </Grid>
  ));