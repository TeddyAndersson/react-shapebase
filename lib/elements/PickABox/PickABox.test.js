import React from 'react';
import PickABox from './PickABox'
import renderer from 'react-test-renderer';

const lorem = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."

test('should render PickABox', () => {
  const component = renderer.create(
    <PickABox
      onClick={()=>{}}
      options={[
        {
          id: '1',
          name: 'Man',
          value: '1',
          label: 'Man',
          icon: 'mars',
        },
        {
          id: '2',
          name: 'Kvinna',
          value: '2',
          label: 'Kvinna',
          icon: 'venus',
        },
        {
          id: '3',
          name: 'Annat',
          value: '3',
          label: 'Annat',
          icon: 'transgender',
        },
      ]}
    />  );
  let tree = component.toJSON();

  expect(tree).toMatchSnapshot();
})
