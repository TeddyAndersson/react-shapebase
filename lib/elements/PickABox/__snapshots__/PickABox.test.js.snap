// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PickABox 1`] = `
<ul
  className="sc-htpNat cRByQz"
>
  <li
    onClick={[Function]}
  >
    <div
      className="sc-bwzfXH cRQsEN sc-bdVaJa jmjnGv"
    >
      <input
        autoComplete="off"
        name="Man"
        type="hidden"
        value="1"
      />
      <div>
        <i>
           A 
        </i>
      </div>
      <div>
        <span>
          Man
        </span>
      </div>
    </div>
  </li>
  <li
    onClick={[Function]}
  >
    <div
      className="sc-bwzfXH cRQsEN sc-bdVaJa jmjnGv"
    >
      <input
        autoComplete="off"
        name="Kvinna"
        type="hidden"
        value="2"
      />
      <div>
        <i>
           A 
        </i>
      </div>
      <div>
        <span>
          Kvinna
        </span>
      </div>
    </div>
  </li>
  <li
    onClick={[Function]}
  >
    <div
      className="sc-bwzfXH cRQsEN sc-bdVaJa jmjnGv"
    >
      <input
        autoComplete="off"
        name="Annat"
        type="hidden"
        value="3"
      />
      <div>
        <i>
           A 
        </i>
      </div>
      <div>
        <span>
          Annat
        </span>
      </div>
    </div>
  </li>
</ul>
`;
